controllers/resource_reconciler.go

```diff
@@ -142,57 +97,121 @@ func (r *ResourceReconciler) Pause(ctx context.Context) error {
+	}
+
+	unobservedTypeResources, observedTypeResources, deletedResources := partitionResources(ownedKinds, r.resources, ownedObjs)
+
+	// Newly-defined types in the CAPZ spec are first recorded in the annotation without performing a
+	// patch that would create resources of that type. CAPZ only patches resources whose kinds have
+	// already been recorded to ensure no resources are orphaned.
+	for _, spec := range unobservedTypeResources {
+		newResourceStatuses = append(newResourceStatuses, infrav1alpha.ResourceStatus{
+			Resource: statusResource(spec),
+			Ready:    false,
+		})
+	}
+
+	for _, spec := range observedTypeResources {
+		spec.SetNamespace(r.owner.GetNamespace())
+
+		if err := controllerutil.SetControllerReference(r.owner, spec, r.Scheme()); err != nil {
+			return false, fmt.Errorf("failed to set owner reference: %w", err)
+		}
+
+		toWatch := meta.AsPartialObjectMetadata(spec)
+		toWatch.APIVersion = spec.GetAPIVersion()
+		toWatch.Kind = spec.GetKind()
+		if err := r.watcher.Watch(log, toWatch, handler.EnqueueRequestForOwner(r.Client.Scheme(), r.Client.RESTMapper(), r.owner)); err != nil {
+			return false, fmt.Errorf("failed to watch resource: %w", err)
+		}
+
+		gvk := spec.GroupVersionKind()
+		log.V(4).Info("applying resource", "resource", klog.KObj(spec), "resourceVersion", gvk.GroupVersion(), "resourceKind", gvk.Kind)
+		err := r.Patch(ctx, spec, client.Apply, client.FieldOwner("capz-manager"), client.ForceOwnership)
+		if err != nil {
+			return false, fmt.Errorf("failed to apply resource: %w", err)
+		}
+
+		ready, err := readyStatus(ctx, spec)
 		if err != nil {
-			return err
+			return false, fmt.Errorf("failed to get ready status: %w", err)
+		}
+		newResourceStatuses = append(newResourceStatuses, infrav1alpha.ResourceStatus{
+			Resource: statusResource(spec),
+			Ready:    ready,
+		})
+	}
+
+	for _, obj := range deletedResources {
+		newStatus, err := r.deleteResource(ctx, obj)
+		if err != nil {
+			return false, fmt.Errorf("failed to delete %s %s/%s", obj.GroupVersionKind(), obj.Namespace, obj.Name)
 		}
 		if newStatus != nil {
 			newResourceStatuses = append(newResourceStatuses, *newStatus)
 		}
 	}
 
+	newOwnedKinds := []metav1.TypeMeta{}
+	for _, status := range newResourceStatuses {
+		typeMeta := metav1.TypeMeta{
+			APIVersion: metav1.GroupVersion{Group: status.Resource.Group, Version: status.Resource.Version}.String(),
+			Kind:       status.Resource.Kind,
+		}
+		if !slices.Contains(newOwnedKinds, typeMeta) {
+			newOwnedKinds = append(newOwnedKinds, typeMeta)
+		}
+	}
+	annotations := r.owner.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	annotations[ownedKindsAnnotation] = getOwnedKindsValue(newOwnedKinds)
+	if annotations[ownedKindsAnnotation] == "" {
+		delete(annotations, ownedKindsAnnotation)
+	}
+	r.owner.SetAnnotations(annotations)
+
 	r.owner.SetResourceStatuses(newResourceStatuses)
 
-	return nil
+	return len(unobservedTypeResources) > 0, nil
```

Comment:
Are we only going to trigger a requeue if `len(unobservedTypeResources) > 0` ? Should we be triggering a requeue nonetheless ?